

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>API Reference &mdash; pandas-genomics 1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="pandas_genomics.io.from_plink" href="io/pandas_genomics.io.from_plink.html" />
    <link rel="prev" title="Documentation for pandas-genomics" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.12
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference external" href="https://halllab.github.io/">Hall Lab Homepage</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/HallLab/pandas-genomics">Pandas-Genomics Github Repo</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-output">Input/Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="io/pandas_genomics.io.from_plink.html">pandas_genomics.io.from_plink</a></li>
<li class="toctree-l3"><a class="reference internal" href="io/pandas_genomics.io.to_plink.html">pandas_genomics.io.to_plink</a></li>
<li class="toctree-l3"><a class="reference internal" href="io/pandas_genomics.io.from_vcf.html">pandas_genomics.io.from_vcf</a></li>
<li class="toctree-l3"><a class="reference internal" href="io/pandas_genomics.io.from_bed.html">pandas_genomics.io.from_bed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#scalars">Scalars</a><ul>
<li class="toctree-l3"><a class="reference internal" href="scalars/pandas_genomics.scalars.Variant.html">pandas_genomics.scalars.Variant</a></li>
<li class="toctree-l3"><a class="reference internal" href="scalars/pandas_genomics.scalars.Genotype.html">pandas_genomics.scalars.Genotype</a></li>
<li class="toctree-l3"><a class="reference internal" href="scalars/pandas_genomics.scalars.Region.html">pandas_genomics.scalars.Region</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#simulation">Simulation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="sim/pandas_genomics.sim.BAMS.html">pandas_genomics.sim.BAMS</a></li>
<li class="toctree-l3"><a class="reference internal" href="sim/pandas_genomics.sim.SNPEffectEncodings.html">pandas_genomics.sim.SNPEffectEncodings</a></li>
<li class="toctree-l3"><a class="reference internal" href="sim/pandas_genomics.sim.PenetranceTables.html">pandas_genomics.sim.PenetranceTables</a></li>
<li class="toctree-l3"><a class="reference internal" href="sim/pandas_genomics.sim.generate_random_gt.html">pandas_genomics.sim.generate_random_gt</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#arrays">Arrays</a><ul>
<li class="toctree-l3"><a class="reference internal" href="arrays/pandas_genomics.arrays.GenotypeDtype.html">pandas_genomics.arrays.GenotypeDtype</a></li>
<li class="toctree-l3"><a class="reference internal" href="arrays/pandas_genomics.arrays.GenotypeArray.html">pandas_genomics.arrays.GenotypeArray</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#accessors">Accessors</a><ul>
<li class="toctree-l3"><a class="reference internal" href="accessors/pandas_genomics.accessors.GenotypeSeriesAccessor.html">pandas_genomics.accessors.GenotypeSeriesAccessor</a></li>
<li class="toctree-l3"><a class="reference internal" href="accessors/pandas_genomics.accessors.GenotypeDataframeAccessor.html">pandas_genomics.accessors.GenotypeDataframeAccessor</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notes.html">Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="release-history.html">Release History</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pandas-genomics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>API Reference</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="module-pandas_genomics.io">
<span id="api-reference"></span><h1>API Reference<a class="headerlink" href="#module-pandas_genomics.io" title="Permalink to this headline">¶</a></h1>
<section id="input-output">
<h2>Input/Output<a class="headerlink" href="#input-output" title="Permalink to this headline">¶</a></h2>
<p>The <cite>io</cite> module provides functions for loading and saving GenotypeArrays or scalar values to common variant formats</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="io/pandas_genomics.io.from_plink.html#pandas_genomics.io.from_plink" title="pandas_genomics.io.from_plink"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_plink</span></code></a>(input[, swap_alleles, ...])</p></td>
<td><p>Load genetic data from plink v1 files (.bed, .bim, and .fam) into a DataFrame.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="io/pandas_genomics.io.to_plink.html#pandas_genomics.io.to_plink" title="pandas_genomics.io.to_plink"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_plink</span></code></a>(data, output[, phenotype_name, ...])</p></td>
<td><p>Save genetic data to plink v1 files (.bed, .bim, and .fam)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="io/pandas_genomics.io.from_vcf.html#pandas_genomics.io.from_vcf" title="pandas_genomics.io.from_vcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_vcf</span></code></a>(filename[, min_qual, drop_filtered])</p></td>
<td><p>Load genetic data from a VCF or BCF file into a DataFrame</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="io/pandas_genomics.io.from_bed.html#pandas_genomics.io.from_bed" title="pandas_genomics.io.from_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_bed</span></code></a>(filename)</p></td>
<td><p>Yields genomic regions from a bed file as Region scalars</p></td>
</tr>
</tbody>
</table>
</section>
<hr class="docutils" />
<span class="target" id="module-pandas_genomics.scalars"></span><section id="scalars">
<h2>Scalars<a class="headerlink" href="#scalars" title="Permalink to this headline">¶</a></h2>
<p>This module contains scalar types, some of which are used in the ExtensionArrays.  They may also be useful on their own.</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="scalars/pandas_genomics.scalars.Variant.html#pandas_genomics.scalars.Variant" title="pandas_genomics.scalars.Variant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Variant</span></code></a>([chromosome, position, id, ref, ...])</p></td>
<td><p>Information about a variant.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="scalars/pandas_genomics.scalars.Genotype.html#pandas_genomics.scalars.Genotype" title="pandas_genomics.scalars.Genotype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Genotype</span></code></a>(variant[, allele_idxs, score])</p></td>
<td><p>Genotype information associated with a specific variant.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="scalars/pandas_genomics.scalars.Region.html#pandas_genomics.scalars.Region" title="pandas_genomics.scalars.Region"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Region</span></code></a>(chromosome, start, end[, name])</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<hr class="docutils" />
<span class="target" id="module-pandas_genomics.sim"></span><section id="simulation">
<h2>Simulation<a class="headerlink" href="#simulation" title="Permalink to this headline">¶</a></h2>
<p>The <cite>sim</cite> module provides classes for generating simulated genotypes</p>
<blockquote>
<div><table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="sim/pandas_genomics.sim.BAMS.html#pandas_genomics.sim.BAMS" title="pandas_genomics.sim.BAMS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BAMS</span></code></a>([pen_table, penetrance_base, ...])</p></td>
<td><p>Biallelic Model Simulator.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="sim/pandas_genomics.sim.SNPEffectEncodings.html#pandas_genomics.sim.SNPEffectEncodings" title="pandas_genomics.sim.SNPEffectEncodings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SNPEffectEncodings</span></code></a>(value)</p></td>
<td><p>Enum: Normalized SNP Effects encoded as 3-length tuples</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="sim/pandas_genomics.sim.PenetranceTables.html#pandas_genomics.sim.PenetranceTables" title="pandas_genomics.sim.PenetranceTables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PenetranceTables</span></code></a>(value)</p></td>
<td><p>Enum: Penetrance Tables for Simple Models</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="sim/pandas_genomics.sim.generate_random_gt.html#pandas_genomics.sim.generate_random_gt" title="pandas_genomics.sim.generate_random_gt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate_random_gt</span></code></a>(variant, alt_allele_freq)</p></td>
<td><p>Simulate random genotypes according to the provided allele frequencies</p></td>
</tr>
</tbody>
</table>
</div></blockquote>
</section>
<hr class="docutils" />
<span class="target" id="module-pandas_genomics.arrays"></span><section id="arrays">
<h2>Arrays<a class="headerlink" href="#arrays" title="Permalink to this headline">¶</a></h2>
<p>This module contains ExtensionArrays and their corresponding ExtensionDtypes</p>
<blockquote>
<div><table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="arrays/pandas_genomics.arrays.GenotypeDtype.html#pandas_genomics.arrays.GenotypeDtype" title="pandas_genomics.arrays.GenotypeDtype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GenotypeDtype</span></code></a>([variant])</p></td>
<td><p>An ExtensionDtype for genotype data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="arrays/pandas_genomics.arrays.GenotypeArray.html#pandas_genomics.arrays.GenotypeArray" title="pandas_genomics.arrays.GenotypeArray"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GenotypeArray</span></code></a>(values[, dtype, copy])</p></td>
<td><p>Holder for genotypes</p></td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Specialized methods are added to the GenotypeArray using Mixins:</p>
<blockquote>
<div><table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">encoding_mixin.EncodingMixin</span></code>()</p></td>
<td><p>Genotype Mixin containing functions for performing encoding</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">info_mixin.InfoMixin</span></code>()</p></td>
<td><p>Genotype Mixin containing functions for calculating various information</p></td>
</tr>
</tbody>
</table>
</div></blockquote>
</section>
<hr class="docutils" />
<span class="target" id="module-pandas_genomics.accessors"></span><section id="accessors">
<h2>Accessors<a class="headerlink" href="#accessors" title="Permalink to this headline">¶</a></h2>
<p>This module contains ‘genomics’ accessors for DataFrames and Series</p>
<blockquote>
<div><table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="accessors/pandas_genomics.accessors.GenotypeSeriesAccessor.html#pandas_genomics.accessors.GenotypeSeriesAccessor" title="pandas_genomics.accessors.GenotypeSeriesAccessor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GenotypeSeriesAccessor</span></code></a>(obj)</p></td>
<td><p>Series accessor for GenotypeArray methods</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="accessors/pandas_genomics.accessors.GenotypeDataframeAccessor.html#pandas_genomics.accessors.GenotypeDataframeAccessor" title="pandas_genomics.accessors.GenotypeDataframeAccessor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GenotypeDataframeAccessor</span></code></a>(pandas_obj)</p></td>
<td><p>DataFrame accessor for GenotypeArray methods</p></td>
</tr>
</tbody>
</table>
</div></blockquote>
</section>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="io/pandas_genomics.io.from_plink.html" class="btn btn-neutral float-right" title="pandas_genomics.io.from_plink" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="index.html" class="btn btn-neutral float-left" title="Documentation for pandas-genomics" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, John McGuigan.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>